From 859990a0b0a3694ee99db398a0ec513e4a02d5de Mon Sep 17 00:00:00 2001
From: Viktor Ashirov <vashirov@redhat.com>
Date: Fri, 28 Nov 2025 12:19:05 +0100
Subject: [PATCH] ipa-pwd-extop: fix memory leaks

In `ipapwd_set_extradata` free `xdata` after it's not longer needed.  It
was leaked because `slapi_value_new_berval()` makes a copy of the data.

In `ipapwd_free_slapi_value_array` free `svals` (caller's pointer)
instead of `sv` (local pointer).

Related: https://pagure.io/freeipa/issue/9895
Signed-off-by: Viktor Ashirov <vashirov@redhat.com>
Reviewed-By: Rob Crittenden <rcritten@redhat.com>
Reviewed-By: Alexander Bokovoy <abbra@users.noreply.github.com>
---
 daemons/ipa-slapi-plugins/ipa-pwd-extop/common.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/daemons/ipa-slapi-plugins/ipa-pwd-extop/common.c b/daemons/ipa-slapi-plugins/ipa-pwd-extop/common.c
index 114d20417d053ad7e822bd474eedf794b2c316d6..de60d4e7bf1b8f0e313e5883041d3e28c09022b9 100644
--- a/daemons/ipa-slapi-plugins/ipa-pwd-extop/common.c
+++ b/daemons/ipa-slapi-plugins/ipa-pwd-extop/common.c
@@ -1107,6 +1107,7 @@ int ipapwd_set_extradata(const char *dn,
 
     slapi_value_free(&va[0]);
     slapi_mods_free(&smods);
+    free(xdata);
 
     return ret;
 }
@@ -1121,7 +1122,7 @@ void ipapwd_free_slapi_value_array(Slapi_Value ***svals)
         }
     }
 
-    slapi_ch_free((void **)sv);
+    slapi_ch_free((void **)svals);
 }
 
 void free_ipapwd_krbcfg(struct ipapwd_krbcfg **cfg)
-- 
2.52.0

